Add setting to automatically show submit button

glolichen opened the pull request:
- Adds a toggle under Developer that automatically shows submit button. Off by default
addresses #199

Gold872 commented on lib/widgets/nt_widgets/single_topic/text_display.dart:
```diff
@@ -38,9 +38,14 @@ class TextDisplayModel extends SingleTopicNTWidgetModel {
-    showSubmitButton ??= ntConnection.getTopicFromName(topic)?.isPersistent;
-    showSubmitButton ??= preferences.getBool(PrefKeys.autoSubmitButton);
-    showSubmitButton ??= false;
+
+    if (preferences.getBool(PrefKeys.autoSubmitButton) ?? false) {
+      showSubmitButton = true;
+    }
+    else {
+      showSubmitButton ??= ntConnection.getTopicFromName(topic)?.isPersistent;
+      showSubmitButton ??= false;
+    }
```
I don't think these lines are necessary, what was there previously should be fine

Gold872 replied:
One issue I found with this is that if `showSubmitButton` is explicity set to false, it will be flipped back to true. Instead, it should be initialized to true only if there was no value present initially:
```dart
showSubmitButton ??= preferences.getBool(PrefKeys.autoSubmitButton);
showSubmitButton ??= ntConnection.getTopicFromName(topic)?.isPersistent;
showSubmitButton ??= false;
```

glolichen replied:
But now, if the `preferences.getBool(PrefKeys.autoSubmitButton)` setting is false, but the topic is persistent, `showSubmitButton` is set to false but it really should be true.

I think these work (I could be wrong):
```dart
    if (preferences.getBool(PrefKeys.autoSubmitButton) ?? false) {
      showSubmitButton ??= true;
    }
    else {
      showSubmitButton ??= ntConnection.getTopicFromName(topic)?.isPersistent;
      showSubmitButton ??= false;
    }
```
so if it has been explicitly set nothing changes, or (very ugly):
```dart
    showSubmitButton ??= (
      (preferences.getBool(PrefKeys.autoSubmitButton) ?? false) ||
      (ntConnection.getTopicFromName(topic)?.isPersistent ?? false)
    );
```

Gold872 commented:
Thanks for making this PR!

Could you add unit tests for these features? There should be one for the settings dialog (in test/widgets/settings_dialog_test.dart), and one for the text display (you'll have to set a mock initial value in shared preferences)

I also think this button could be moved out of developer settings since I wouldn't say it's a very advanced feature, the appearance settings should work just fine for this

glolichen commented:
Moved the setting under appearance settings; added unit tests for text display dialog and settings page. I tried to follow what the other tests did but I am not sure these are the kind you wanted. I also tested the new setting on a real robot and it works.

I think the changes made earlier in `text_display.dart` are necessary to automatically show the submit button based on the SharedPreferences. Deleting it breaks the new unit tests.

Gold872 commented on lib/pages/dashboard_page.dart:
```diff
@@ -1657,6 +1657,10 @@ class _DashboardPageState extends State<DashboardPage> with WindowListener {
           setState(() {});
         },
+        onAutoSubmitButtonChanged: (value) async {
+          preferences.setBool(PrefKeys.autoSubmitButton, value);
```
You should add an `await` before this line

Gold872 commented on lib/services/settings.dart:
```diff
@@ -74,4 +74,5 @@ class PrefKeys {
   static String logLevel = 'log_level';
   static String gridDpiOverride = 'grid_dpi_override';
   static String windowPosition = 'window_position';
+  static String autoSubmitButton = 'auto_submit_button';
```
Maybe change this to `auto_text_submit_button`?

Gold872 commented on lib/widgets/settings_dialog.dart:
```diff
@@ -51,6 +51,7 @@ class SettingsDialog extends StatefulWidget {
   final void Function(Level? level)? onLogLevelChanged;
   final FutureOr<void> Function(String? value)? onGridDPIChanged;
   final void Function()? onOpenAssetsFolderPressed;
+  final void Function(bool value)? onAutoSubmitButtonChanged;
```
Make this a `FutureOr<void>` so that way you can add an `await` statement before calling it, which will only make the checkbox update after the setting has been saved

Gold872 commented on lib/widgets/settings_dialog.dart:
```diff
@@ -434,6 +438,32 @@ class _SettingsDialogState extends State<SettingsDialog> {
+          Flexible(
+            child: DialogToggleSwitch(
+              initialValue: false,
+              label: 'Automatically Show Submit Button',
```
This could be a bit confusing to some users, maybe clarify which submit button this by changing it to "Auto Show Text Submit Button"

Gold872 commented on lib/widgets/settings_dialog.dart:
```diff
@@ -434,6 +438,32 @@ class _SettingsDialogState extends State<SettingsDialog> {
+            child: DialogToggleSwitch(
+              initialValue: false,
+              label: 'Automatically Show Submit Button',
+              onToggle: (value) {
+                setState(() {
+                  widget.onAutoSubmitButtonChanged?.call(value);
+                });
```
See my comment above on the `FutureOr<void>` but after adding that, this should be changed to:
```dart
onToggle: (value) async {
  await widget.onAutoSubmitButtonChanged?.call(value);
  setState(() {});
}
```

Gold872 commented on test/widgets/nt_widgets/single-topic/text_display_test.dart:
```diff
@@ -28,7 +29,7 @@ void main() {
 
   setUp(() async {
-    SharedPreferences.setMockInitialValues({});
+    SharedPreferences.setMockInitialValues({ PrefKeys.autoSubmitButton: false });
```
I would recommend not setting the submit button preference for every run, as there's only a few tests that would actually be using it

Gold872 commented on test/widgets/nt_widgets/single-topic/text_display_test.dart:
```diff
@@ -82,7 +83,10 @@ void main() {
   group('Show submit button defaults to', () {
-    test('true if topic is persistent', () {
+    test('true if topic is persistent and auto submit button true', () {
+      preferences.setBool(PrefKeys.autoSubmitButton, true);
+      expect(preferences.getBool(PrefKeys.autoSubmitButton), isTrue);
```
Instead of setting the preference here and testing it, it would instead be better to just do:
```dart
SharedPreferences.setMockInitialValues({PrefKeys.autoTextSubmitButton: true});
preferences = await SharedPreferences.getInstance();
```
Which is done in several of the tests that rely on different preferences

Gold872 commented on test/widgets/nt_widgets/single-topic/text_display_test.dart:
My advice for the tests is to leave the current ones as is, and instead add one or two additional ones that test it with the shared preferences, otherwise this could make the tests very complicated (which was an issue at one point in the past)

Gold872 commented on test/widgets/settings_dialog_test.dart:
```diff
@@ -883,4 +886,51 @@ void main() {
+          body: SettingsDialog(
+            preferences: preferences,
+            ntConnection: createMockOfflineNT4(),
+            onAutoSubmitButtonChanged: (value) {
+              fakeSettings.changeAutoSubmitButton();
+              preferences.setBool(PrefKeys.autoSubmitButton, value);
+            },
```
I would make this async and add an `await` before setting the preference

Some of the other settings tests do this

Gold872 reviewed (changes requested):
Amazing work on this! There's just a few issues with the unit tests, variable naming, and some minor async things, but the UI to this look good.

Also, the CI is failing since the formatting is incorrect, to fix this, run these 3 commands:
`dart format .`
`dart run import_sorter:main`
`dart fix --apply`

glolichen commented:
Thanks for reviewing!

I reverted to the old "true if topic is persistent" test without setting the preference, and added two tests for preference is true and preference is false.

I do still have a question on what [text_display.dart](https://github.com/Gold872/elastic-dashboard/pull/200/files/3ffe6ea6c2a9ae1dfe3b63ac0617eff3ae03d23e..6b99a375ffdebe01a806f5cdbb516a91a74b816c#diff-892fa5ecf8e630489c91c8263fe6a6267350bfdb3e624580d7464ce731c71d14) should actually be, I put my proposed new code in the commit but please see if there is anything wrong.

Gold872 commented:
> I do still have a question on what [text_display.dart](https://github.com/Gold872/elastic-dashboard/pull/200/files/3ffe6ea6c2a9ae1dfe3b63ac0617eff3ae03d23e..6b99a375ffdebe01a806f5cdbb516a91a74b816c#diff-892fa5ecf8e630489c91c8263fe6a6267350bfdb3e624580d7464ce731c71d14) should actually be, I put my proposed new code in the commit but please see if there is anything wrong.

I think the way it is now is good, I realize now the reason for it, which is the situation where the topic is persistent but the preference is set to false.

Gold872 reviewed (approved):
This looks great!

What I'd like to add for 2026 is the ability to set default settings for any widget. I'll have to think more about the implementation of it, but for the time being this will be a useful addition.
